gax/retry_loop.h

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gax/backoff_policy.h"
+#include "gax/call_context.h"
+#include "gax/retry_policy.h"
+#include "gax/status.h"
+#include <chrono>
+#include <functional>
+#include <thread>
+#include <type_traits>
+
+#ifndef GAPIC_GENERATOR_CPP_GAX_RETRY_LOOP_H_
+#define GAPIC_GENERATOR_CPP_GAX_RETRY_LOOP_H_
+
+namespace google {
+namespace gax {
+
+template <typename RequestT, typename ResponseT>
+gax::Status MakeRetryCall(
+    CallContext& context, RequestT const& request, ResponseT* response,
+    std::function<Status(CallContext&, RequestT const&, ResponseT*)> next_stub,
+    std::unique_ptr<RetryPolicy> retry_policy,
```

Comment:
We discussed it offline, but I'm still not fully convinced that we can't just use regular pointers here. It should be safe and will simplify testing. Please either change it to regular pointers or at least provide a comment describing why smart pointer by value parameter is peered here.

Reply:
The caller could do something like 
````
auto retry_clone = clone_retry(context);
auto backoff_clone = clone_backoff(context);
return MakeRetryCall(context, request, response, std::move(make_retry), retry_clone.get(), backoff_clone.get());
```
Is that what you mean? We have to do that sometimes when we don't control method signatures, but in this case we do.


Herb Sutter talks about smart ptrs in method signatures here: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/ . For the most part he agrees with you, but there are a couple of reasons I wanted to use unique_ptr instead of const& or raw ptr.

1) It makes it easier to reuse (or modify) code for asynchronous operations. If async code receives references and then starts running on a different thread or in a deferred manner, the reference may dangle. Ownership is unclear in all cases when the parameter is a raw ptr.
2) It simplifies the retry loop. During a previous iteration I had a complicated multi-parameter check-and-clone thing in the retry loop in order to pass both the default (owned by the RetryStub) and the client's version (injected by the user during construction).